### ISSUE TITLE
Trigger a sync immediately when a doc changes locally

### ISSUE THREAD

**Issue by garethbowen** [medic/cht-core#5738]
When a doc changes locally we update the UI to show a sync is required and then wait for the 5 minute timer to expire before syncing. Instead when we detect a change we should immediately attempt to sync unless we know the user is offline. This will improve the average sync time a little.

**Comment by garethbowen** [comment]
Ready for AT in `5738-sync-immediately`

**Comment by kennsippell** [comment]
LG is currently struggling to stay on the `C5.4xlarge`, CPU-bound by _changes requests. This would potentially generate a lot more server load. Do we have an idea of how this will affect server health for production instances?

**Comment by garethbowen** [comment]
It'll definitely have an effect but hopefully not much. We could consider separating the upwards replication from the downwards replication so we can sync data up frequently (gentle on the server) and sync down less often (the `_changes` requests).

What if we kept the current replication scheme for syncing up, but reduced syncing down to either a) clicking the sync button manually, or b) 30 minutes since last sync down?

**Comment by benkags** [comment]
The experience for 2-way texting using the messaging tab was affected us syncing at an interval of 5 mins. An offline user using messaging for chat has to wait for up to 5 minutes for the next sync before a message can be sent or received. That or manually sync every time they want a message sent or to receive a message. The desired behaviour would be that messages sync immediately if the user has internet connectivity.

I was going to open an issue on this but it looks related to this. Do you think fixing this would cater for this as well? @garethbowen Or do I open a separate issue for this?

**Comment by benkags** [comment]
Here is another scenarios that may potentially be solved by this:

If a report submission is configured to trigger an immediate SMS notification, this notification cannot go out until the syncing has happened. This is particularly relevant to both 2-way-texting projects in Zimbabwe and South Africa where a nurse uses the automated message sent to the client to verify that the client gave a valid phone number before confirming a valid enrollment.

cc @Maigua

**Comment by garethbowen** [comment]
This issue will improve that situation somewhat because any message or report generated on the user's device will cause an automatic sync attempt, so assuming they're online the report will sync immediately. So this will help one way of the two-way texting use case, and solve the SMS generated from report submission use case.

Changes originating on the server will still take up to five minutes to replicate down to the phone. Please raise a new issue if this is a problem. I'm considering making the server to client replication interval configurable - this would mean for projects that need high responsiveness it could be set to 1 minute or so, but for projects that would prefer greater scalability it could be set to 30 minutes or so. What do you think?

**Comment by benkags** [comment]
I was thinking along the same lines. I think making it configurable will work. I will go ahead and raise that issue.

**Comment by kennsippell** [comment]
It's great we have a long-term plan for this feature! What is the plan in 3.9? 

Are we thinking to merge this as-written?  I suspect this change risks blocking some partners from upgrading to 3.9 (concretely LG-Uganda).

**Comment by garethbowen** [comment]
> Are we thinking to merge this as-written?

Yes. Why do you think this will block some partners? I'm assuming due to server load? This won't account for much additional load. The practical difference is it changes _when_ the replication happens to be more immediate rather than _how often_ it happens.

**Comment by kennsippell** [comment]
Yes. I am concerned about server load [as stated here](https://github.com/medic/cht-core/issues/5738#issuecomment-603101376). LG-UG has repeated expressed concerns.

Can this change result in a user running multiple automated replications per 5 minute period? Let's say a user sits and makes 5 new contacts for a family and that takes them 5 minutes - won't that result in 5 replications in 5 minutes for that user? 

> This won't account for much additional load. 
What are you doing to understand the impact this change will have more clearly. Can you share your thinking here?

Have you engaged LG to get feedback on this change?

**Comment by garethbowen** [comment]
> Can this change result in a user running multiple automated syncs in a 5 minute period? Let's say a user registers a new family - so they make 5 new contacts in 5 minutes - will that result in 5 syncs in 5 minutes for that user?

Yes that's correct. In that situation the number of requests will go from around 2 to around 5.

> Can you share your analysis here more clearly? What is this based on and what do you mean by "much"?

Because of the way replication works the first request in either scenario will be the most expensive so adding around 3 more in the next few minutes won't increase the load on the server much. Given the performance improvements we've made with the changes endpoint in recent releases, for any project on 3.7 or earlier (and particularly 3.6.2 and earlier) I expect the upgrade to be an overall win for server load, with more improvements to come in subsequent releases.

For more context, this improvement was raised as a requirement to ensure that the data on the server is up to date as much as possible and is to rectify what is seen by some as an oversight in the initial periodic syncing implementation.

**Comment by SCdF** [comment]
@rmugwanya can provide more detail, but this ticket is important for itech zim

**Comment by rmugwanya** [comment]
In the case of VMMC (I-Tech Zim), when an enrollment happens, it is a requirement that the enrollment confirmation message/sms is received by the client promptly before the leave the enrollment point either at the facility or the remote site. Essentially the enrollment Nurse has to confirm with the client that the message has come through to their phone. It is important that this message comes through quickly enough before the client proceeds to access other services i.e. HIV Testing, Counseling, The VMMC Surgery, etc. 

Once an enrollment is done, a task is generated to initiate the sending of client check-in sms. However, You have to do a manual sync for this task to appear. Can we have a more seamless process where user does not have to do manual syncs
CC @benkags @Maigua

**Comment by SCdF** [comment]
Any work on this ticket should be charged to: `117 I-TECH-Zimbabwe`

**Comment by dianabarsan** [comment]
Assigning myself for AT.

**Comment by dianabarsan** [comment]
Looking good. I'm seeing a sync happening every time I change something. 

However, If I edit something while a sync is in progress, the change doesn't trigger a sync and when the current sync is finished, my sync status says I'm up to date but my latest change is yet to be uploaded. 

Don't know if we should cover this in this issue or separately.

**Comment by garethbowen** [comment]
I think this is covered by #6299

**Comment by garethbowen** [comment]
Merged.
